### PR TITLE
feat: gc for zombie piece & metaTask & bucket migration

### PR DESCRIPTION
**Pull request by jingjunLi**
### Description

- GCZombiePiece
GCZombiePieceTask is an abstract interface to record the information for collecting the piece store space by deleting **zombie pieces** data that dues to any exception, the piece data meta is not on chain but the pieces have been stored in piece store.

- GCMeta
GCMetaTask is an abstract interface to record the information for collecting the SP meta store space by deleting the expired data.

- GC for Bucket Migration
When bucket migration is completed or has failed, we need to delete redundant data to free up space:
  1. If the bucket migration is successful, we delete the data from the source node.
  2. If the bucket migration fails, we delete the data from the dest node

### Implementation of the GCZombie

- New Configurations
    - `GCConfig::EnableGCZombie`: Enables or disables the GCZombie feature.
    - `GCConfig::GCZombiePieceTimeInterval`: Time interval for generating GCZombie tasks (default: `DefaultGlobalBatchGcZombiePieceTimeInterval` - 10*60 seconds).
    - `ParallelConfig::GlobalGCZombieParallel`: Maximum allowed parallel GCZombie tasks (default: 1).
    - `GCZombieSafeObjectIDDistance`: A reserve of object IDs during GCZombie deletion. If the scanned object ID range plus `DefaultGlobalGcZombieSafeObjectIDDistance` exceeds the current maximum system ID, the scan resumes from 0.
    - `GCZombiePieceObjectIDInterval`: Interval between generated object IDs for each GCZombie task. For example, task 1 might handle IDs 0-100, task 2 101-200, and so on (default: `DefaultGlobalGcZombiePieceObjectIDInterval` - 100).

Core Processes

image

- Generation of `GCZombiePieceTask`: Triggered by `gcZombiePieceTicker` resulting in the creation of GCZombiePieceTask.
- Processing of `GCZombiePieceTask`: `ExecuteModular::HandleGCZombiePieceTask`.
    - `gcZombiePieceFromIntegrityMeta`: Determines whether a piece is a ZombiePiece based on the `IntegrityMeta` table. Scans all `IntegrityMeta` within the current object ID range specified in GCZombiePieceTask (`StartObjectId`, `EndObjectId`).
    - `gcZombiePieceFromPieceHash`: Determines whether a piece is a ZombiePiece based on the `PieceHash` table. Scans all `PieceHash` within the current object ID range specified in GCZombiePieceTask (`StartObjectId`, `EndObjectId`).
- Report Handling of `GCZombiePieceTask`: `ManageModular::HandleGCZombiePieceTask`.
    - If the task is successful, it is removed from the `gcZombieQueue`.
    - If the task fails, it is reinserted into the queue. If the retry limit is exceeded, the task is canceled.

###  Implementation of the Meta GC

Mainly for the GC (Garbage Collection) of different GC meta, currently periodically deleting expired data from two tables: `bucketTraffic` and `readRecord`.

New Configurations:

- Expiration Time and Interval Configuration:
    - `bucketTrafficKeepLatestDay`: Configured in `ExecutorConfig::BucketTrafficKeepTimeDay`, defaults to `DefaultExecutorBucketTrafficKeepTimeDay` (retaining the latest 180 days).
    - `readRecordKeepLatestDay`: Configured in `ExecutorConfig::ReadRecordKeepTimeDay`, defaults to `DefaultExecutorReadRecordKeepTimeDay` (retaining the latest 30 days).
    - `DefaultGlobalGCMetaTimeInterval`: Configures the time interval for the generation of GC Meta Tasks, set to 10 * 60 (equivalent to 10 minutes).
    - `ReadRecordDeleteLimit`: Maximum number of records to be deleted from the read record table in each gc meta task. (default 100)
- Functional Configuration:
    - `GCConfig::EnableGCMeta`: Controls whether GCMeta is enabled or not.
    - `SQLDBConfig::EnableTracePutEvent`: Controls whether Trace Put Event is enabled or not.

Core Process:
image

1. Generation of GCMetaTask:
Triggered by `gcMetaTicker`, resulting in the creation of `GCMetaTask`.
2. Processing of GCMetaTask:
Handled in `ExecuteModular::HandleGCMetaTask`
    - `gcMetaBucketTraffic`: Deletes entries from `BucketTraffic` using `SpDBImpl::DeleteAllBucketTrafficExpired` for expired `BucketTrafficTable`.
    - `gcMetaReadRecord`: Deletes entries from `ReadRecord` using `SpDBImpl::DeleteAllReadRecordExpired` for expired `ReadRecord` table.

### Implementation of the BucketMigration GC

**Purpose:**

1. After successful bucket migration, delete data from the source node.
2. In case of failed or canceled bucket migration, delete migrated data from the destination node.

**New Configurations:**

- GCBucketMigrationTask-related settings and parameters:
    - Two new parameters:
        - `gcBucketMigrationTimeout` (`MinGCBucketMigrationTime` **0.5-1 hour**): Timeout duration for the task.
        - `gcBucketMigrationRetry` (`MinGCBucketMigrationRetry` **3-5 times**): Retry attempts for the task.
    - Task Type: `GfSpGCBucketMigrationTask`.
    - Task scheduling priority: `TypeTaskGCZombiePiece`. Adjusted priority:
        - `TypeTaskGCMeta` *→* `DefaultSmallerPriority / 4`
        - `TypeTaskGCBucketMigration` *→* `DefaultSmallerPriority / 4`

**Core Workflow:**

- Generation of GCBucketMigrationTask:
    - For successful migration (source node deletes old data):
        1. Destination SP notifies source SP of migration completion: `GfSpClient::NotifyPostMigrateBucket` → `GfSpNotifyPostMigrate` → `ManageModular::NotifyPostMigrateBucket`.
        2. Upon successful migration, `ManageModular::GenerateGCBucketMigrationTask` generates a `GCBucketMigrationTask`.
        3. Enqueues the task in `ManageModular::gcBucketMigrationQueue`.
    - For failed migration (destination node deletes migrated data):
        - `BucketMigrateScheduler`::`PostMigrateBucket` generates a `GCBucketMigrationTask` on the destination node if migration fails.
- Execution of GCBucketMigrationTask:
    - `ExecuteModular::HandleGCBucketMigrationBucket`.
    - Iterates through all GVGs of the bucket and retrieves objects within each GVG using the `ListObjectsByGVGAndBucketForGC` interface.
    - Checks if the location information of each object is valid. If a segment should not be on this SP, it is cleaned.
- Report Handling of GCBucketMigrationTask:
    - If the task is successful, it is removed from `gcBucketMigrationQueue`.
    - If the task fails, it is reinserted into the queue. If the retry limit is exceeded, the task is canceled.

### Changes

Notable changes: 
* add each change in a bullet point here
* ...

### CODE REVIEW

File: store/sqldb/traffic.go [hunk 1]
```diff
@@ -451,6 +451,15 @@ func (s *SpDBImpl) UpdateBucketTraffic(bucketID uint64, update *corespdb.BucketT
 	return nil
 }
 
+// DeleteAllBucketTrafficExpired update the bucket traffic in traffic db with the new traffic
+func (s *SpDBImpl) DeleteAllBucketTrafficExpired(yearMonth string) (err error) {
+	result := s.db.Where("month < ?", yearMonth).Delete(&BucketTrafficTable{})
```
Comment (flywukong) [inline store/sqldb/traffic.go:456]: need to lock the table when delete?

File: store/sqldb/traffic.go [hunk 2]
```diff
@@ -630,3 +639,12 @@ func (s *SpDBImpl) GetUserReadRecord(userAddress string, timeRange *corespdb.Tra
 	}
 	return records, nil
 }
+
+// DeleteAllReadRecordExpired delete all read record before ts(ts is UnixMicro)
+func (s *SpDBImpl) DeleteAllReadRecordExpired(ts uint64) (err error) {
+	result := s.db.Where("read_timestamp_us < ?", ts).Delete(&ReadRecordTable{})
```
Comment (flywukong) [inline store/sqldb/traffic.go:645]: may needs to lock the table when delete

File: store/sqldb/traffic.go (hunk 1 above)
Comment (flywukong) [reply store/sqldb/traffic.go:456]: Restrict users from deleting this month’s bucket traffic

File: store/sqldb/object_integrity.go [hunk 3]
```diff
@@ -199,6 +205,78 @@ func (s *SpDBImpl) DeleteObjectIntegrity(objectID uint64, redundancyIndex int32)
 	return err
 }
 
+type ByRedundancyIndexAndObjectID []*corespdb.IntegrityMeta
+
+func (a ByRedundancyIndexAndObjectID) Len() int { return len(a) }
```
Comment (sysvm) [inline store/sqldb/object_integrity.go:210]: rename `a` to `b`?

File: modular/manager/manage_task.go [hunk 4]
```diff
@@ -40,6 +40,8 @@ var (
 	ErrNotifyMigrateSwapOut = gfsperrors.Register(module.ManageModularName, http.StatusNotAcceptable, 60006, "failed to notify swap out start")
 )
 
+const BucketMigrationGCWaitTime = 10 * time.Second
```
Comment (sysvm) [inline modular/manager/manage_task.go:43]: This constant is not exported, use private.

**Review by sysvm — APPROVED** [review]
LGTM

**Review by flywukong — APPROVED** [review]

File: modular/executor/execute_gc.go [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+		log.CtxErrorw(ctx, "failed to query replicate piece checksum", "task_info", task.Info(), "error", err)
+		return err
+	}
+
+	if len(waitingVerifyGCPieces) == 0 {
+		log.CtxErrorw(ctx, "no waiting gc pieces", "task_info", task.Info())
+		return nil
+	}
+
+	for _, piece := range waitingVerifyGCPieces {
```
Comment (alexgao001) [inline modular/executor/execute_gc.go:517]: a object might have multiple piece hash records, seems the call to chain or meta are duplicate.

File: store/sqldb/object_integrity.go [hunk 6]
```diff
@@ -199,6 +205,78 @@ func (s *SpDBImpl) DeleteObjectIntegrity(objectID uint64, redundancyIndex int32)
...
+				Where("object_id >= ? and object_id <= ?", startObjectID, endObjectID).
+				Limit(ListObjectsDefaultSize).
+				Order("object_id,redundancy_index asc").
+				Find(&objects).Error
+			totalObjects = append(totalObjects, objects...)
+		}
+	} else {
+		for i := 0; i < IntegrityMetasNumberOfShards; i++ {
+			integrityMetasTableName := GetIntegrityMetasTableNameByShardNumber(i)
+			joins := fmt.Sprintf("right join buckets on buckets.bucket_id = %s.bucket_id", integrityMetasTableName)
```
Comment (alexgao001) [inline store/sqldb/object_integrity.go:254]: this join might take times, maybe subquery from each tables first is better.

File: store/sqldb/object_integrity.go [hunk 7]
```diff
@@ -199,6 +205,78 @@ func (s *SpDBImpl) DeleteObjectIntegrity(objectID uint64, redundancyIndex int32)
...
+			metrics.SPDBTime.WithLabelValues(SPDBFailureGetObjectIntegrity).Observe(
+				time.Since(startTime).Seconds())
+		} else {
+			metrics.SPDBCounter.WithLabelValues(SPDBSuccessGetObjectIntegrity).Inc()
+			metrics.SPDBTime.WithLabelValues(SPDBSuccessGetObjectIntegrity).Observe(
+				time.Since(startTime).Seconds())
+		}
+	}()
+
+	if includePrivate {
```
Comment (ruojunm) [inline store/sqldb/object_integrity.go:241]: maybe we don't have to distinguish private or public objects here, as SP need gc for all zombie pieces.

**Comment by ruojunm** [comment]
LGTM

**Review by ruojunm — APPROVED** [review]

**Review by alexgao001 — APPROVED** [review]

**Review by will-2012 — APPROVED** [review]

**Pull request merged**